### DISCUSSION TITLE
How to correctly organize code for orm so that whether a model is active or not is explictly stated

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#10051] [Usage Questions] [answered]
How to correctly organize code for orm so that whether a model is active or not is explictly stated.

To elaborate on my question,
When I fetch some models from database, there are two scenarios:
1. The model will be passed into other functions as arguments, and these functions will update the model, in the end, the outer function will commit.
For example,
```
def outer_func():
   sess = get_session()
    models = get_from_db(sess):
    inner_func(models). # models will be updated
    commit_to_db(sess)
```
2. The model will be read from database, it is passed to inner func to do some computation; inner func will not update it.

My question is, when writing the `inner_func`, how can I make sure the future developer later read the `inner_func` understands if the model is intended to be updated or not.

The only thing I can do atm is to explicitly name the argument, e.g., `ro_model`, `w_model`. I also used to manually convert all the ro_model into pydantic models and pass into the functions. But this is kinda verbose as I have to dup all the fields of a Model Class into a Pydantic model. (Note that, sqlmodel is not my choice since the maintenance on the project is not active).

Is there something better?

**Comment by CaselIT** [comment]
Hi,

The question is not very clear to me. Let me try anyway.
Session can expunge some models so that changes are no longer traked. So maybe that's wat you want for your read only models?

**Reply by ghost** [reply]
Yes. I'd like to know:
1. How to expunge the tracking
2. How this expunge is conveyed so that for a function taking a model, I know it has been expunged.

**Comment by zzzeek** [comment]
do you want to use decorators of some kind for the "RO" version such that if any changes occur within the session, it throws an error?  that's the best I can think of

**Reply by ghost** [reply]
That sounds great.

**Reply by ghost** [reply]
I feel like this is a common request since people are passing models around in the system. There shall be a way to tell if the mode is active in a trasaction or not.

**Reply by zzzeek** [reply]
these are framework features and SQLAlchemy is not a framework.   The Session should provide everything you need (typically event hooks and some introspection features) that you can use to tell when things happen.   

in this case, "Active in a transaction" does not indicate changes or not. All SQL including SELECT statements are emitted within transactions.  SELECT statements do change the state of the database (try SELECTING from a table in PostgreSQL, then while leaving psql open, try dropping the table in another psql session).  so you'd want to use some other criteria for this.  depends on what kinds of state you are trying to detect.    Mostly, the easiest way would be to disallow calls to .commit() a transaction.

**Reply by ghost** [reply]
Is there any example on the decorator you mentioned?

**Reply by zzzeek** [reply]
this is a proof of concept, it depends very much on how different functions get access to the `Session` and what specific operations you want to guard against.  guarding against `commit` is one option here, you could also guard against any objects entering the `Session.dirty`, `Session.new` or `Session.deleted` collections also

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import sessionmaker


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

e = create_engine("sqlite://")

get_session = sessionmaker(e)

@event.listens_for(get_session, "before_commit")
def _guard_commits(sess):
    if sess.info.get("_guard_commits", False):
        raise Exception("no commits allowed")



def ro_method(fn):
    def go(sess, *arg):
        existing_guard_commit = sess.info.get("_guard_commits", False)
        sess.info["_guard_commits"] = True
        try:
            return fn(sess, *arg)
        finally:
            sess.info["_guard_commits"] = existing_guard_commit
    return go

@ro_method
def inner_func(sess, models):
    # calling this will raise
    # sess.commit()

    return models

def get_from_db(sess):
    return sess.scalars(select(A)).one()

def commit_to_db(sess):
    sess.commit()

def outer_func(e):
    sess = get_session()
    models = get_from_db(sess)
    inner_func(sess, models)
    commit_to_db(sess)


if __name__ == "__main__":
    with e.begin() as conn:
        Base.metadata.create_all(conn)

    with get_session() as sess:
        a1 = A(data='a1')
        sess.add(a1)
        sess.commit()

    outer_func(e)
```

**Reply by ghost** [reply]
Thanks. The above approach requires passing sessoin to the inner func. Is it possible to get the session from the model ?

**Reply by zzzeek** [reply, marked as answer]
yes if "model" is an ORM object you can do this:

```py
from sqlalchemy import inspect
session = inspect(obj).session
```
